cloudfoundry/managers/bits/bitsmanager.go

```diff
@@ -321,11 +321,32 @@ func (m BitsManager) RetrieveZip(path string) (ZipFile, error) {
 				baseName = params["filename"]
 			}
 		}
-		return ZipFile{
-			r:        resp.Body,
-			baseName: baseName,
-			filesize: fileSize,
-		}, nil
+		if fileSize > 0 {
+			return ZipFile{
+				r:        resp.Body,
+				baseName: baseName,
+				filesize: fileSize,
+			}, nil
+		} else {
+			tempFile, err := ioutil.TempFile("", "")
+			if err != nil {
+				return ZipFile{}, err
+			}
+			defer os.Remove(tempFile.Name())
```

Comment:
Wondering if you'll run into a race condition here i.e. as soon as the function returns the file would be removed so the `ZipFile{}` struct will refer to a deleted file. Perhaps better to read the file content in memory set a Reader? (it could potentially balloon memory usage, but it might be the most robust option)

Reply:
I was also concerned about this, but then checked what the implementation of os.Remove() is: https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/os/file_unix.go;l=295

As far as I can see it's mapped to the syscall unlink(): https://man7.org/linux/man-pages/man2/unlink.2.html

> If the name was the last link to a file but any processes still have the file open, the file will remain in existence until the last file descriptor referring to it is closed.

The windows implementation works similarly: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-deletefilea

Which seems pretty safe. Another way would be, as you suggest, to use a memory buffer. I'm just a little more afraid of a large file filling up memory than a temp file.

Reply:
@oysteing have you tested the PR and confirmed the `unlink()` is not an issue?